### PR TITLE
testing examples added

### LINKED ISSUE #2550: Document the custom `sender` & `value` feature for remix-tests
See https://github.com/ethereum/remix/pull/1402

### LINKED ISSUE #2636: Update unit testing docs for natspec
Update unit testing docs to make sure it has the info described here: https://medium.com/remix-ide/solidity-unit-testing-using-remix-tests-part-2-50a9f486ab5d

### PR DESCRIPTION
**Pull request by Aniket-Engg**
- fixes #2550 
- fixes #2636 
- Added testing examples

### CODE REVIEW

File: docs/unittesting_examples.md [hunk 1]
```diff
@@ -1,4 +1,318 @@
...
+            // by zero, etc. inside getData.
+            Assert.ok(false, 'failed unexpected');
+        }
+    }
+    
+    /// Test failure case of 'add' using try-catch
+    function testAddFailureUsingTryCatch1() public {
+        // This will revert on 'require(class > 0 && class <= 12, "Invalid class");' for class '13'
+        try ar.add(101, 'secondStudent', 13) returns (uint256 r) {
+            Assert.ok(false, 'method execution should fail');
```
Comment (ryestew) [inline docs/unittesting_examples.md:205]: I'll ask you about this line tomorrow - about seeing the 'method execution should fail'

File: docs/unittesting.md [hunk 2]
```diff
@@ -39,6 +39,45 @@ Click the button "Run tests" to executes all tests whose box has been checked be
 
 ![](images/a-unit-testing-run-result.png)
 
+Customization
+------------------
+Remix facilitates users with various types of customizations to test a contract properly.
+
+**1. Custom Compiler Context**
+
+`Solidity Unit Testing` refers `Solidity Compiler` plugin for compiler configurations. One can provide customized inputs for `Compiler`, `EVM Version` & `Enable Optimization`.
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting.md:48]: updated

File: docs/unittesting.md [hunk 3]
```diff
@@ -39,6 +39,45 @@ Click the button "Run tests" to executes all tests whose box has been checked be
...
+
+![](images/a-unit-testing-custom-compiler-config.png)
+
+**2. Custom Transaction Context**
+
+For a contract method interaction, prime parameters of transaction are `from` address, `value` & `gas`. Usually, we need to test method's behaviour under different values of these parameters.
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting.md:54]: done

File: docs/unittesting.md [hunk 4]
```diff
@@ -39,6 +39,45 @@ Click the button "Run tests" to executes all tests whose box has been checked be
...
+    Assert.equal(msg.sender, TestsAccounts.getAccount(0), "wrong sender in checkSenderIs0AndValueis10");
+    Assert.equal(msg.value, 10, "wrong value in checkSenderIs0AndValueis10");
+}
+```
+
+**Note:** To use custom `msg.sender` functionality, `remix_accounts.sol` should be imported in your test file. 
+
+Complete example can be seen in [examples](./unittesting_examples) section.
+
+Regarding `gas`, Remix estimate the required gas for each transaction internally. Still if a contract deployment fails with `Out-of-Gas` error, it tries to redeploy it by doubling the gas. Deployment failing with double gas will show error: `contract deployment failed after trying twice: The contract code couldn\'t be stored, please check your gas limit`
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting.md:71]: done, but with 3 backticks too, it is going out of column

File: docs/unittesting.md [hunk 5]
```diff
@@ -39,6 +39,45 @@ Click the button "Run tests" to executes all tests whose box has been checked be
...
+
+
+Points to remember
+------------------
+
+* A test contract can not have a method with parameters. Having one such method will show error: `Method 'methodname' can not have parameters inside a test contract`
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting.md:77]: done

File: docs/unittesting.md [hunk 6]
```diff
@@ -39,6 +39,45 @@ Click the button "Run tests" to executes all tests whose box has been checked be
...
+* No. of test accounts are `3` before remix-ide release v0.10.0 and `10` afterwards
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting.md:78]: done

File: docs/unittesting.md [hunk 7]
```diff
@@ -39,6 +39,45 @@ Click the button "Run tests" to executes all tests whose box has been checked be
...
+* A test file which imports `remix_accounts.sol` might not compile successfully with `Solidity Compiler` plugin but it will work fine with Solidity Unit Testing plugin
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting.md:79]: Done

File: docs/unittesting_examples.md [hunk 8]
```diff
@@ -1,4 +1,318 @@
...
+        updateOwner(account2);
+        // check if owner is set to expected account i.e. account2
+        Assert.equal(getOwner(), account2, 'owner should be updated to account2');
+    }
+}
+```
+
+### 3. Testing method execution
+
+With Solidity, one can verify the changes made by a method in storage by accessing those variables out of a contract. But while testing whether method execution was successful and if execution failed, what was the reason behind it, can also be an importnat case.
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting_examples.md:136]: updated

File: docs/unittesting_examples.md [hunk 9]
```diff
@@ -1,4 +1,318 @@
...
+```
+pragma solidity >=0.4.22 <0.7.0;
+contract AttendanceRegister {
+struct Student{
+        string name;
+        uint class;
+    }
+event Added(string name, uint class, uint time);
+mapping(uint => Student) public register; // roll number => student details
+    function add(uint rollNumber, string memory name, uint class) public returns (uint256){
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting_examples.md:150]: yes. It is fixed.

File: docs/unittesting_examples.md (hunk 1 above)
Comment (Aniket-Engg) [reply docs/unittesting_examples.md:205]: That Assert is symbolic. That line will be never executed with `ar.add(101, 'secondStudent', 13)` throwing error. If due to any mistake in providing input for fail case, it is somehow passed then that `Assert` will fail stating `method execution should fail`.
Let me know your thoughts.

File: docs/unittesting_examples.md [hunk 10]
```diff
@@ -1,4 +1,318 @@
...
+    }
+
+    /// Test if value is set as expected
+    function valueIsSet200() public returns (bool) {
+        foo.set(200);
+        return Assert.equal(foo.get(), 200, "value is not 200");
+    }
+}
+```
+
```
[the remark this answers is not in the document]
Comment (Aniket-Engg) [reply docs/unittesting_examples.md:59]: I have updated the contract variables. Added and updated some comment to make it better explanatory. Please have a look @ryestew

**Pull request merged**